services/lgtm/lgtm-grade.service.js

```diff
@@ -1,99 +1,88 @@
 'use strict'
 
-const LegacyService = require('../legacy-service')
-const { makeBadgeData: getBadgeData } = require('../../lib/badge-data')
-const { checkErrorResponse } = require('../../lib/error-helper')
-
-// This legacy service should be rewritten to use e.g. BaseJsonService.
-//
-// Tips for rewriting:
-// https://github.com/badges/shields/blob/master/doc/rewriting-services.md
-//
-// Do not base new services on this code.
-module.exports = class LgtmGrade extends LegacyService {
-  static get category() {
-    return 'analysis'
-  }
+const LgtmBaseService = require('./lgtm-base')
 
+module.exports = class LgtmGrade extends LgtmBaseService {
   static get route() {
     return {
       base: 'lgtm/grade',
+      pattern: ':language/g/:user/:repo',
+    }
+  }
+
+  async handle({ language, user, repo }) {
+    const data = await this.fetch({ user, repo })
+
+    if (!('languages' in data)) {
+      return this.constructor.render({
+        grade: 'invalid response data',
+        color: 'red',
+      })
+    }
+
+    const languageLabel = (() => {
+      switch (language) {
+        case 'cpp':
+          return 'c/c++'
+        case 'csharp':
+          return 'c#'
+        // Javascript analysis on LGTM also includes TypeScript
+        case 'javascript':
+          return 'js/ts'
+        default:
+          return language
+      }
+    })()
+
+    let grade = 'no language data'
+    let color = 'red'
+
+    for (const languageData of data.languages) {
+      if (languageData.lang === language && 'grade' in languageData) {
+        // Pretty label for the language
+        grade = languageData.grade
+        // Pick colour based on grade
+        if (languageData.grade === 'A+') {
+          color = 'brightgreen'
+        } else if (languageData.grade === 'A') {
+          color = 'green'
+        } else if (languageData.grade === 'B') {
+          color = 'yellowgreen'
+        } else if (languageData.grade === 'C') {
+          color = 'yellow'
+        } else if (languageData.grade === 'D') {
+          color = 'orange'
+        }
+      }
+    }
```

Comment:
Can we split these 2 blocks for generating the colour and label out into helper functions. Again, calling from `render()`.